test/src/userobjects/InterfaceDomainUserObject.C

```diff
@@ -0,0 +1,164 @@
+  _robin_bnd_ids = std::set<BoundaryID>(robin_bnd_ids_vec.begin(), robin_bnd_ids_vec.end());
+
+  const auto & interface_boundaries = getParam<std::vector<BoundaryName>>("interface_boundaries");
+  const auto & interface_bnd_ids_vec = _mesh.getBoundaryIDs(interface_boundaries);
+  _interface_bnd_ids =
+      std::set<BoundaryID>(interface_bnd_ids_vec.begin(), interface_bnd_ids_vec.end());
+
+  for (const auto interface_bnd_id : _interface_bnd_ids)
+  {
+    const auto & interface_connected_blocks = _mesh.getInterfaceConnectedBlocks(interface_bnd_id);
+    for (const auto interface_connected_block : interface_connected_blocks)
+    {
+      if (hasBlocks(interface_connected_block))
+        // we're operating on this block
+        continue;
+
+      // these are blocks connected to our blocks
+      _interface_connected_blocks.insert(interface_connected_block);
+    }
+  }
+}
+
+void
+InterfaceDomainUserObject::initialize()
+{
+  _nodal_integrals.clear();
+  _nodal_integrals.resize(_subproblem.mesh().getMesh().max_node_id());
+}
+
+void
+InterfaceDomainUserObject::executeOnElement()
+{
+  for (const auto i : make_range(_test.size()))
+  {
+    auto & integral = _nodal_integrals[_current_elem->node_ref(i).id()];
+    for (const auto qp : make_range(qRule().n_points()))
+      integral += JxW()[qp] * coord()[qp] *
+                  (_grad_u[qp] * _grad_test[i][qp] - _test[i][qp] * _func.value(_t, qPoints()[qp]));
+  }
+}
+
+void
+InterfaceDomainUserObject::executeOnBoundary()
+{
+  if (!_robin_bnd_ids.count(_current_boundary_id))
+    return;
+
+  for (const auto i : make_range(_test_face.size()))
+  {
+    auto & integral = _nodal_integrals[_current_elem->node_ref(i).id()];
+    for (const auto qp : make_range(qRule().n_points()))
+      integral += JxW()[qp] * coord()[qp] * _test_face[i][qp] * _robin_coef * _u[qp];
+  }
+}
+
+void
+InterfaceDomainUserObject::executeOnInterface()
+{
+  if (!_interface_bnd_ids.count(_current_boundary_id))
+    return;
+
+  for (const auto i : make_range(_test_face.size()))
+  {
+    auto & integral = _nodal_integrals[_current_elem->node_ref(i).id()];
+    for (const auto qp : make_range(qRule().n_points()))
+      integral += JxW()[qp] * coord()[qp] * _test_face[i][qp] * _interface_penalty *
+                  (_u[qp] - _v_neighbor[qp]);
+  }
+}
+
+void
+InterfaceDomainUserObject::finalize()
+{
+  gatherSum(_nodal_integrals);
+  for (const auto integral : _nodal_integrals)
+    if (std::abs(integral) > _nl_abs_tol)
+      mooseError("Interface diffusion problem not converged");
+}
+
+void
+InterfaceDomainUserObject::threadJoin(const UserObject & y)
+{
+  const InterfaceDomainUserObject & interface_uo =
+      dynamic_cast<const InterfaceDomainUserObject &>(y);
+  for (unsigned int i = 0; i < _nodal_integrals.size(); i++)
+    _nodal_integrals[i] += interface_uo._nodal_integrals[i];
+}
+
+void
+InterfaceDomainUserObject::checkVariable(const MooseVariableFieldBase & variable) const
```

Comment:
With my other suggestions, we can move this into `DomainUserObject`. I like to have this because letting derived class implement this does not seem ideal.